aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreError.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.datastore;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.amplifyframework.api.graphql.GraphQLResponse;
+import com.amplifyframework.core.model.Model;
+
+import java.util.Objects;
+
+/**
+ * Contains error object from GraphQL server response from
+ * an attempt to publish a mutation. This object will be passed
+ * back to the user via {@link DataStoreErrorHandler} to be
+ * appropriately handled by custom logic.
+ * @param <M> type of model that caused publication error.
+ */
+public final class DataStoreError<M extends Model> {
+    private final GraphQLResponse.Error error;
+    private final M local;
+    private final M remote;
+
+    /**
+     * Constructs a new error instance containing details regarding
+     * a failed publication to the cloud and involved models.
+     * @param error the error response from the server
+     * @param local the local model pending mutation
+     * @param remote the remote model if applicable
```

Comment:
We should bottom out on this. Right now, the `DataStoreError` looks _alot_ like the `ConflictData` bundle used by the conflict handler. I think the two concepts are getting increasingly close to eachother.

Reply:
I think it's _fine_.

As I mentioned before, I tried consolidating all of these into generic AppSync related objects, and frankly, it wasn't super difficult since they contain similar data. However, I'm not sure if introducing breaking changes at this point is a good idea or even worth the effort. We already make it clear that `ConflictUnhandled` error is handled specifically by the conflict handler, so a little bit of separation might make it more clear what we are doing internally.

Reply:
We should aim to have our error handler look like the JS and iOS ones. If we're going to do something different, let's loop them in.

Reply:
Exposing `remote` model was my attempt to mirror the JS error handler input. The rest of the parameters are exposed in the form of `GraphQLResponse.Error`, because the user should be able to obtain the errorType and such from it. I thought about exposing it directly, but I didn't do it in case that the GraphQL error was not a direct AppSync error. 

Speaking of which, my current implementation doesn't pass back the operation type! I will add that in before I merge the changes.